posts/2018-06-07-shared-state-in-fp.md

```diff
@@ -0,0 +1,372 @@
+---
+layout: post
+title: Shared State in Functional Programming
+
+meta:
+  nav: blog
+  author: gvolpe
+  pygments: true
+
+tut:
+  scala: 2.12.4
+  binaryScala: "2.12"
+  dependencies:
+    - org.scala-lang:scala-library:2.12.4
+    - org.typelevel::cats-core:1.1.0
+
+---
+
+Newcomers to Functional Programming are often very confused about the proper way to share state without breaking purity and end up having a mix of pure and impure code that defeats the purpose of having pure FP code in the first place.
+
+This is the reason that has motivated me to write a beginner friendly guide :)
+
+## Referential Transparency
+
+I think this is the key to understanding the benefits of Functional Programming. Without it, it'll be hard to understand other deeper concepts, such as `Shared State`. So let's start this guide by refreshing our knowledge about it:
+
+##### What is Referential Transparency?
+
+Being able to substitute an expression by its bound value without changing the meaning of the program.
+
+##### What does this mean?
+
+An example is worth thousand words. Consider this:
+
+```scala
+val expr = 123
+(expr, expr)
+```
+
+If we now substitute the expression by its bound value, does the meaning of the program change?
+
+```
+(123, 123)
+```
+
+No! The program remains the same. We call this property *Referential Transparency*. How about this?
+
+```scala
+val expr = println("hey")
+(expr, expr)
+```
+
+Let's perform susbtitution once again:
+
+```scala
+(println("hey"), println("hey"))
+```
+
+M'kay... The meaning of the program has now changed. The first program prints "hey" once whereas the second program prints "hey" twice. This is not `Referential Transparent`.
+
+So having this knowledge, let's now move onto the next section!
+
+## Use Case
+
+We have a program that runs three computations at the same time and updates the internal state to keep track of the
+tasks that have been completed. When all the tasks are completed we request the final state and print it out.
+
+You should get an output similar to the following one:
+
+```
+Starting process #1
+Starting process #2
+Starting process #3
+  ... 3 seconds
+Done #2
+  ... 5 seconds
+Done #1
+  ... 10 seconds
+Done #3
+List(#2, #1, #3)
+```
+
+We'll use the concurrency primitive `Ref[IO, List[String]]` to represent our internal state because it's a great fit.
+
+### Getting started
+
+So this is how we might decide to start writing our code:
+
+```scala
```

Comment:
Consider using `tut` here so it actually goes through the type checker in CI http://tpolecat.github.io/tut/

Going to stop writing this for the rest of the review, but there's a couple code blocks below that can also benefit